src/nodes.coffee

```diff
@@ -2065,20 +2096,20 @@ exports.For = class For extends While
         {index: ivar, name, @step, isComplex: isComplexOrAssignable}
     else
       svar    = @source.compile o, LEVEL_LIST
-      if (name or @own) and not IDENTIFIER.test svar
+      if (name or @own) and @source.unwrap() not instanceof IdentifierLiteral
         defPart    += "#{@tab}#{ref = scope.freeVariable 'ref'} = #{svar};\n"
         svar       = ref
       if name and not @pattern
         namePart   = "#{name} = #{svar}[#{kvar}]"
       if not @object
         defPart += "#{@tab}#{step};\n" if step isnt stepVar
-        lvar = scope.freeVariable 'len' unless @step and stepNum and down = (parseNum(stepNum[0]) < 0)
+        lvar = scope.freeVariable 'len' unless @step and stepNum? and down = (stepNum < 0)
```

Comment:
Doing an assignment in this condition is really disgusting. I know you didn't design it, but while we're here can you restructure it?